### PR TITLE
Implement ZSink.collectAllToMapN

### PR DESCRIPTION
**Pull request by mijicd**
Part of #4886

### CODE REVIEW

File: streams-tests/shared/src/test/scala/zio/stream/experimental/ZSinkSpec.scala [hunk 1]
```diff
@@ -33,6 +33,16 @@ object ZSinkSpec extends ZIOBaseSpec {
               .run(ZSink.collectAllToMap((_: Int) % 3)(_ + _))
           )(equalTo(Map[Int, Int](0 -> 18, 1 -> 12, 2 -> 15)))
         ),
+        suite("collectAllToMapN")(
```
Comment (iravid) [inline streams-tests/shared/src/test/scala/zio/stream/experimental/ZSinkSpec.scala:36]: Same re tests from ZTransducerSpec - they cover some interesting edge cases

File: streams-tests/shared/src/test/scala/zio/stream/experimental/ZSinkSpec.scala [hunk 2]
```diff
@@ -33,6 +33,29 @@ object ZSinkSpec extends ZIOBaseSpec {
...
+          testM("respects the given limit") {
+            ZStream
+              .fromChunk(Chunk(1, 1, 2, 2, 3, 2, 4, 5))
+              .transduce(ZSink.collectAllToMapN(2)((_: Int) % 3)(_ + _))
+              .runCollect
+              .map(assert(_)(equalTo(Chunk(Map(1 -> 2, 2 -> 4), Map(0 -> 3, 2 -> 2), Map(1 -> 4, 2 -> 5)))))
+          },
+          testM("collects as long as map size doesn't exceed the limit") {
+            ZStream
+              .fromChunks(Chunk(0, 1, 2), Chunk(3, 4, 5), Chunk(6, 7, 8, 9))
```
Comment (mijicd) [inline streams-tests/shared/src/test/scala/zio/stream/experimental/ZSinkSpec.scala:46]: As mentioned in the chat, this test fails. When invoked using `fromChunk(Chunk(0, 1, 2, ..., 9))` the test succeeds. Investigating the cause.

Comment (mijicd) [reply streams-tests/shared/src/test/scala/zio/stream/experimental/ZSinkSpec.scala:46]: Fixed the off by one in decompose / decomposeM.
